submissions/andrewklmn/memory-pair-game/app.js

```diff
@@ -0,0 +1,189 @@
+const files = [
+  'image1.png',
+  'image2.png',
+  'image3.png',
+  'image4.png',
+  'image5.png',
+  'image6.png',
+  'image7.png',
+  'image8.png',
+];
+
+const imageNames = [...files,...files];
+const cardsShuffler = function() { return 0.5 - Math.random() };
+const gameBoard = document.querySelector(".gameboard");
+const controlDiv = document.querySelector('.control');
+
+let numberOfFails = 0; 
+
+const openAllCards = function() {
+  const cards = gameBoard.querySelectorAll('.flip-container');
+  cards.forEach((c,i)=>{
+    setTimeout(() => {
+      c.classList.remove('guessed');
+      c.classList.add('opened');      
+    }, 100*i);
+  })
+};
+
+const closeAllCards = function() {
+  const cards = gameBoard.querySelectorAll('.flip-container');
+  cards.forEach((c,i)=>{
+    setTimeout(() => {
+      c.classList.remove('guessed');
+      c.classList.remove('opened');     
+    }, 100*i);
+  })
+};
+
+const initBoard = function () {
+
+  gameBoard.innerHTML = '';
+
+  imageNames.forEach((image)=>{
+    
+    const fragment = document.createDocumentFragment();
+
+    const flipContainer = document.createElement('div');
+    flipContainer.classList.add('flip-container')
+    flipContainer.id = image;
+
+    const flipper = document.createElement('div');
+    flipper.classList.add('flipper');
+
+    const front = document.createElement('div'); 
+    front.classList.add('front'); 
+    
+    const frontImage = document.createElement('img'); 
+    frontImage.classList.add('card');
+    frontImage.src = 'img/empty.png'; 
+    frontImage.alt = 'Back';
+
+    front.appendChild(frontImage);
+
+    const back = document.createElement('div'); 
+    back.classList.add('back');
+
+    const backImage = document.createElement('img'); 
+    backImage.classList.add('card');
+    backImage.src = `img/${image}`; 
+    backImage.alt = 'Image';
+    
+    back.appendChild(backImage);
+    flipper.appendChild(front);
+    flipper.appendChild(back);
+    flipContainer.appendChild(flipper);
+    fragment.appendChild(flipContainer);
+
+    gameBoard.appendChild(fragment);
+  });
+
+}
+
+const spreadCards = function () {
+  
+  const backImages = gameBoard.querySelectorAll('.back > img');
+
+  imageNames.sort(cardsShuffler);
+
+  backImages.forEach((image,i) => {
+    image.src = `img/${imageNames[i]}`;
+    image.parentNode.parentNode.parentNode.id = imageNames[i];
+  });
```

Comment:
In my opinion it would be better to shuffle the cards before drawing and not do double work.

Reply:
> In my opinion it would be better to shuffle the cards before drawing and not do double work.

@madmaxWMFU  You are right.  Shuffle function at the beginning can be called twice.
The first time in the initialization function, and the second time in the Start button listener.

There are two ways to start a new game after page opening:
1) you can just click any closed card and the game will be started immediately without preview.
2) you can press the Start button with the same result, but with (or without) a preview before the game.

So, cards must be already shuffled in the first case... 
That is why I put the shuffling call in the initialization function.

It is needed only for the first new game which can be started just by clicking any of the closed cards.
But, after this first game, cards will be shuffled only when the Start button will be pressed just before a new game.

Reply:
It seems I understand what you mean... )))